File: Binary_Insertion_Sort/Binary_Insertion_Sort.php
<?php

function binarySearch($arr,$h,$item){
	$l = 0;
	while($l<$h){
		$m = floor(($l+$h)/2);
		if($arr[$m] == $item){
			return $m+1;
		}
		elseif ($arr[$m] > $item) {
			$h = $m - 1;
		}elseif($arr[$m] < $item){
			$l = $m + 1;
		}
	}

	return ($item > $arr[$l])?$l+1:$l;
}

function binaryInsertionSort($arr){
	for ($i=0; $i <count($arr); $i++) { 
		$selected = $arr[$i];
		$j = $i;
		$location = binarySearch($arr,$j-1,$selected);
		while ($j>=1 && $j>$location) {
			$arr[$j] = $arr[$j-1];
			$j--;
		}
		$arr[$j] = $selected;
	}

	return $arr;
}

function printArray($arr){
	$output="[";
	for ($i=0; $i < count($arr); $i++) { 
		$output.=" ".$arr[$i]." ";
		if ($i<count($arr)-1) {
			$output.=",";
		}
	}
	$output.="]";

	echo $output;
}

$n = (int)readline();
$arr = array();

for ($i=0; $i < $n; $i++) { 
	$arr[$i] = (int)readline();
}

$arr = binaryInsertionSort($arr);
printArray($arr);

?>

File: Binary_Insertion_Sort/Binary_Insertion_Sort.R
n = as.integer(readline())
arr = c()

for (i in 1:n) {
	arr[i] = as.integer(readline())
}

binarySearch<-function(arr,h,item){
	l = 1
	while(l<h){
		m = floor((l+h)/2)
		if(arr[m] == item){
			return(m+1)
		}else if(arr[m] > item){
			h = m - 1
		}else if(arr[m] < item){
			l = m + 1
		}
	}

	return(ifelse(arr[l]<item,l+1,l))
}

binaryInsertionSort<-function(arr){

	for (i in 2:length(arr)) {
		selected = arr[i]
		j = i
		location = binarySearch(arr,j-1,selected)
		while(j>=2 && j > location){
			arr[j] = arr[j-1]
			j = j - 1
		}
		arr[j] = selected
	}
  
  return(arr)
}

printArray<-function(arr){
	output = "["

	for (i in 1:length(arr)) {
		output = paste(output," ",arr[i]," ")
		if(i<length(arr)){
			output = paste(output,",")
		}
	}
	output = paste(output,"]")

	print(output)
}

arr = binaryInsertionSort(arr)
printArray(arr)
